cmd/dagger/codegen.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"go/parser"
+	"go/token"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"dagger.io/dagger"
+	"github.com/dagger/dagger/codegen/generator"
+	gogenerator "github.com/dagger/dagger/codegen/generator/go"
+	nodegenerator "github.com/dagger/dagger/codegen/generator/nodejs"
+	"github.com/dagger/dagger/codegen/introspection"
+	"github.com/dagger/dagger/core/envconfig"
+	"github.com/dagger/dagger/engine/client"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+var (
+	codegenFlags  = pflag.NewFlagSet("codegen", pflag.ContinueOnError)
+	codegenOutput string
+	codegenPkg    string
+)
+
+func init() {
+	codegenFlags.StringVarP(&codegenOutput, "output", "o", "", "output file")
+	codegenFlags.StringVar(&codegenPkg, "package", "main", "package name")
+}
+
+func RunCodegen(
+	ctx context.Context,
+	engineClient *client.Client,
+	_ *dagger.Client,
+	envCfg *envconfig.Config,
+	depEnvs []*dagger.Environment,
+	cmd *cobra.Command,
+	_ []string,
+) (rerr error) {
+	if workdir != "" {
+		if err := os.Chdir(workdir); err != nil {
+			return err
+		}
+	}
+
+	pkg, err := getPackage(envCfg.SDK)
+	if err != nil {
+		return err
+	}
+
+	introspectionSchema, err := generator.Introspect(ctx, engineClient)
+	if err != nil {
+		return err
+	}
+
+	generated, err := generate(ctx, introspectionSchema, generator.Config{
+		Package:                pkg,
+		Lang:                   generator.SDKLang(envCfg.SDK),
+		EnvironmentName:        envCfg.Name,
+		DependencyEnvironments: depEnvs,
+	})
+	if err != nil {
+		return err
+	}
+
+	output, err := getOutput(envCfg.SDK)
+	if err != nil {
+		return err
+	}
+
+	if output == "" || output == "-" {
+		cmd.Println(string(generated))
```

Comment:
super nit: maybe consider an early return here, for a moment I thought we were doing an `os.WriteFile("", ...)` down below, before analyzing the indentation